### PR TITLE
Improve the speed of `smoothcal.solve_2d_DPSS` by adding additional filtering options in `nucal._linear_fit`

### PR DESCRIPTION
**Pull request by tyler-a-cox**
This PR improves the speed of `smoothcal.solve_2D_DPSS` by using `nucal._linear_fit` to allow for additional filtering options. For the nominal inputs of `smooth_cal.CalibrationSmoother.time_freq_2D_filter`, the new `lu_solve` method is roughly 12x faster than the previously used `pinv` method.

For other uses of this function such as [full_day_rfi.ipynb](https://github.com/HERA-Team/hera_notebook_templates/blob/master/notebooks/full_day_rfi.ipynb) and [full_data_auto_checker.ipynb](https://github.com/HERA-Team/hera_notebook_templates/blob/master/notebooks/full_day_auto_checker.ipynb) where filtering scale is finer than the typical `smooth_cal` run, the new `lu_solve`  method is significantly faster with comparable results.

Screenshot 2023-04-08 at 3 32 35 PM

This PR also deprecates the `XTXinv` keyword argument in favor of a more general `cached_input`, which also stores intermediate data products computed by `scipy.linalg.lu_factor`.

### CODE REVIEW

File: hera_cal/smooth_cal.py [hunk 1]
```diff
@@ -122,42 +120,53 @@ def solve_2D_DPSS(gains, weights, time_filters, freq_filters, XTXinv=None):
...
+            'pinv' uses np.linalg.pinv to compute the least squares solution and tends to be the most reliable. 'lu_solve'
+            uses scipy.linalg.lu_solve to compute the least squares solution and tends to be the fastest. 
+            'solve' uses np.linalg.solve and 'lstsq' uses np.linalg.lstsq and have comparable results.
+        cached_input: Matrix of (X^T W X)^{-1} for the input DPSS filters and weights. Useful for filtering
             many gain grids with similar flagging patterns. np.ndarray of shape (N_time_vectors * N_freq_vectors,
-            N_time_vectors * N_freq_vectors)
+            N_time_vectors * N_freq_vectors). Can be obtained using the 'cached_output' return value from a
+            previous call to this function.
+        XTXinv: Matrix of (X^T W X)^{-1} for the input DPSS filters and weights. 
+            Warning: This is deprecated and will be removed in a future version. Use cached_input instead.
```
Comment (jsdillon) [inline hera_cal/smooth_cal.py:132]: As far as I know, the only one using the current interface is the notebook... let's just rip off the bandaid and break the API

File: hera_cal/smooth_cal.py [hunk 2]
```diff
@@ -122,42 +120,53 @@ def solve_2D_DPSS(gains, weights, time_filters, freq_filters, XTXinv=None):
             obtained from hera_cal.smooth_cal.dpss_filters
         freq_filters: DPSS filtering vectors along the frequency axis, (Nfreqs, N_freq_vectors)
             obtained from hera_cal.smooth_cal.dpss_filters
-        XTXinv: Matrix of (X^T W X)^{-1} for the input DPSS filters and weights. Useful for filtering
+        method: method to use for solving the linear least squares problem. Options are 'pinv', 'lstsq', 'lu_solve', and 'solve'.
+            'pinv' uses np.linalg.pinv to compute the least squares solution and tends to be the most reliable. 'lu_solve'
+            uses scipy.linalg.lu_solve to compute the least squares solution and tends to be the fastest. 
+            'solve' uses np.linalg.solve and 'lstsq' uses np.linalg.lstsq and have comparable results.
+        cached_input: Matrix of (X^T W X)^{-1} for the input DPSS filters and weights. Useful for filtering
```
Comment (jsdillon) [inline hera_cal/smooth_cal.py:127]: Perhaps you want to describe what the keys are

File: hera_cal/smooth_cal.py [hunk 3]
```diff
@@ -122,42 +120,53 @@ def solve_2D_DPSS(gains, weights, time_filters, freq_filters, XTXinv=None):
...
-
-        # Use einsum to calculate (X^T W X) in a memory efficient way
-        XTX = np.einsum(
-            "ij,kl,jl,jm,ln->ikmn", np.transpose(time_filters), np.transpose(freq_filters),
-            weights, time_filters, freq_filters, optimize=True
+    # Check for deprecated XTXinv argument
+    if XTXinv is not None:
+        warnings.warn(
+            "The XTXinv argument is deprecated and will be removed in a future version. Use cached_input instead.",
+            DeprecationWarning,
```
Comment (jsdillon) [inline hera_cal/smooth_cal.py:144]: You can dump this, I think

File: hera_cal/smooth_cal.py [hunk 4]
```diff
@@ -122,42 +120,53 @@ def solve_2D_DPSS(gains, weights, time_filters, freq_filters, XTXinv=None):
...
-        XTX = np.reshape(XTX, (ncomps, ncomps))
-        XTXinv = np.linalg.pinv(XTX)
-
+    # Make sure data types between weights and filters are compatible for einsum optimization
+    weights = weights.astype(time_filters.dtype)
+
+    # Calculate the number of components in the design matrix
+    ncomps = time_filters.shape[1] * freq_filters.shape[1]
+
+    # Use jax einsum to calculate (X^T W X) in a memory efficient way
```
Comment (jsdillon) [inline hera_cal/smooth_cal.py:152]: In general, I think it's helpful to label the einsum axes in the comments.

File: hera_cal/smooth_cal.py [hunk 5]
```diff
@@ -122,42 +120,53 @@ def solve_2D_DPSS(gains, weights, time_filters, freq_filters, XTXinv=None):
...
+    XTX = jnp.einsum(
+        "ij,kl,jl,jm,ln->ikmn", np.transpose(time_filters), np.transpose(freq_filters),
+        weights, time_filters, freq_filters, optimize=True
+    )
+    XTX = np.reshape(XTX, (ncomps, ncomps))
+        
     # Calculate X^T W y using the property (A \otimes B) vec(y) = (A Y B)
-    XTWy = np.ravel(np.transpose(time_filters) @ (gains * weights) @ freq_filters)
+    XTWy = np.ravel(np.dot(np.dot(np.transpose(time_filters), (gains * weights)), freq_filters))
```
Comment (jsdillon) [inline hera_cal/smooth_cal.py:160]: any reason not to use `jnp`?

File: hera_cal/smooth_cal.py [hunk 6]
```diff
@@ -122,42 +120,53 @@ def solve_2D_DPSS(gains, weights, time_filters, freq_filters, XTXinv=None):
...
-        XTXinv = np.linalg.pinv(XTX)
-
+    # Make sure data types between weights and filters are compatible for einsum optimization
+    weights = weights.astype(time_filters.dtype)
+
+    # Calculate the number of components in the design matrix
+    ncomps = time_filters.shape[1] * freq_filters.shape[1]
+
+    # Use jax einsum to calculate (X^T W X) in a memory efficient way
+    XTX = jnp.einsum(
```
Comment (jsdillon) [inline hera_cal/smooth_cal.py:153]: If we have the cache, is it possible to avoid this calculation? Would that be worthwhile?

File: hera_cal/smooth_cal.py [hunk 7]
```diff
@@ -292,7 +301,7 @@ def time_filter(gains, wgts, times, filter_scale=1800.0, nMirrors=0):
 
 def time_freq_2D_filter(gains, wgts, freqs, times, freq_scale=10.0, time_scale=1800.0,
                         tol=1e-09, filter_mode='rect', maxiter=100, window='tukey', method='CLEAN',
-                        dpss_vectors=None, XTXinv=None, eigenval_cutoff=1e-9, skip_flagged_edges=True,
+                        dpss_vectors=None, fit_method="pinv", cached_input={}, eigenval_cutoff=1e-9, skip_flagged_edges=True,
```
Comment (jsdillon) [inline hera_cal/smooth_cal.py:304]: Need a docstring for cached_input to replace XTXinv

File: hera_cal/smooth_cal.py [hunk 8]
```diff
@@ -960,7 +977,7 @@ def time_freq_2D_filter(self, freq_scale=10.0, time_scale=1800.0, tol=1e-09, fil
 
                 # If the weights grid is the same as the previous, the solution matrix can be reused to speed up computation
                 if method == 'DPSS' or method == 'dpss_leastsq':
-                    XTXinv = info['XTXinv'] if np.allclose(wgts_grid, wgts_old) else None
+                    cached_input = info['cached_input'] if np.allclose(wgts_grid, wgts_old) else {}
```
Comment (jsdillon) [inline hera_cal/smooth_cal.py:980]: This reminds me... I think we might get a speed-up if we sort the gain_grids by the number of flags... so that antennas with the same flagging pattern are more likely to come in a row.

File: hera_cal/nucal.py [hunk 9]
```diff
@@ -708,14 +708,27 @@ def _linear_fit(XTX, Xy, solver='lu_solve', tol=1e-15):
         tol : float, default=1e-15
             Tolerance to use for regularization. If method is 'lu_solve' or 'solve', this is added to the diagonal of XTX. 
             If method is 'pinv' or 'lstsq', this is used as the rcond parameter for np.linalg.pinv and np.linalg.lstsq respectively.
+        cached_input : dictionary, default={}
```
Comment (jsdillon) [inline hera_cal/nucal.py:711]: I'd explicitly reference the text for cached_output to describe the format this should take.

File: hera_cal/nucal.py [hunk 10]
```diff
@@ -724,32 +737,49 @@ def _linear_fit(XTX, Xy, solver='lu_solve', tol=1e-15):
...
-    elif (solver == "lu_solve" or solver == "solve"):
+    elif solver == "lu_solve" or solver == "solve":
         XTX[np.diag_indices_from(XTX)] += tol
 
-    if solver == 'lu_solve':
+    if solver == "lu_solve":
         # Factor XTX using scipy.linalg.lu_factor
-        L = linalg.lu_factor(XTX)
+        if "LU" in cached_input:
+            L = cached_input.get('LU')
```
Comment (jsdillon) [inline hera_cal/nucal.py:746]: Looks like you missed testing using the cache for lu_solve... should be add a test where you just run the smoother again.

**Review by jsdillon — CHANGES_REQUESTED** [review]
Nice work @tyler-a-cox! This is looking pretty good, just a few minor changes (mostly regarding documentation).

File: hera_cal/smooth_cal.py (hunk 8 above)
Comment (tyler-a-cox) [reply hera_cal/smooth_cal.py:980]: Just added a couple lines to sort the antennas by number of flags

File: hera_cal/nucal.py (hunk 10 above)
Comment (tyler-a-cox) [reply hera_cal/nucal.py:746]: I ended up testing this in test_nucal

**Review by jsdillon — APPROVED** [review]
Nice work @tyler-a-cox
